wasmlib/helpers/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+        Mutex::new(m)
+    });
+
+    pub static GLOBAL_WAC_BLOCKSTORE: Lazy<Mutex<HashMap<Cid, &[u8]>>> = Lazy::new(|| {
+        let m = HashMap::new();
+        Mutex::new(m)
+    });
+}
+
+/// Given a pointer to the start of a byte array of WAC data
+/// encode the data into the codec
+///
+///
+/// # Safety
+///
+/// This function assumes that the load_raw_block function returns valid results
+pub unsafe fn load_raw_block_caller(blk_cid: Cid) -> Result<Box<[u8]>, libipld::error::Error> {
+    let blk_cid_bytes = blk_cid.to_bytes();
+    let cidptr = blk_cid_bytes.as_ptr();
+
+    let blk_res = &*load_raw_block(cidptr, blk_cid_bytes.len() as u8);
+    let res = blk_res.to_result();
+    match res {
+        Ok(v) => Ok(v.to_owned().into_boxed_slice()),
+        Err(err) => Err(err),
+    }
+}
+
+/// Given a pointer to the start of a byte array of WAC data
+/// encode the data into the codec
+///
+///
+/// # Safety
+///
+/// This function assumes that the load_raw_block function returns valid results
+pub unsafe fn load_wac_block_caller(blk_cid: Cid) -> Result<Box<[u8]>, libipld::error::Error> {
+    let blk_cid_bytes = blk_cid.to_bytes();
+    let cidptr = blk_cid_bytes.as_ptr();
+
+    let blk_res = &*load_wac_block(cidptr, blk_cid_bytes.len() as u8);
+    let res = blk_res.to_result();
+    match res {
+        Ok(v) => Ok(v.to_owned().into_boxed_slice()),
+        Err(err) => Err(err),
+    }
+}
+
+use std::ffi::c_void;
+
+use libipld::Cid;
+
+pub trait Codec {
+    /// Given a pointer to the start of a byte array of WAC data
+    /// encode the data into the codec
+    ///
+    /// # Safety
+    ///
+    /// This function assumes the block pointer has size have been allocated and filled.
+    unsafe fn encode(&self, ptr: *mut u8, len: u32) -> *const ValueOrError;
+
+    /// Given a pointer to the start of a byte array and
+    /// its length, decode it into a standard IPLD codec representation
+    /// (for now WAC)
+    ///
+    /// # Safety
+    ///
+    /// This function assumes the block pointer has size have been allocated and filled.
+    unsafe fn decode(&self, ptr: *mut u8, len: u32) -> *const ValueOrError;
+}
+
+pub trait AdlBase {
+    /// Takes a pointer and length of a byte array containing WAC encoded data and returns
+    /// a pointer to the ADL instance.
+    ///
+    /// # Safety
+    ///
+    /// This function assumes the block pointer has size have been allocated and filled.
+    unsafe fn load_adl(&self, ptr: *mut u8, len: u32) -> *mut ADLorWAC;
+}
+
+#[allow(clippy::missing_safety_doc)]
+pub trait AdlBytes {
+    unsafe fn new_bytes_reader(&self, adlptr: *const c_void) -> *mut c_void;
+    unsafe fn read_adl(&self, adlptr: *mut u8, bufptr: *mut u8, bufleni: i32) -> *const ReadResp;
+    unsafe fn seek_adl(&self, adlptr: *mut c_void, offset: i64, whence: u32) -> u64;
+}
+
+#[allow(clippy::missing_safety_doc)]
+pub trait AdlMap {
+    unsafe fn new_map_iter(&self, adlptr: *const c_void) -> *mut c_void;
```

Comment:
Could we have "seeking" be part of this? Or having some extra options that could be passed in.

Reply:
Yep, I mostly copied these ADLs to match the ones from go-ipld-prime https://github.com/ipld/go-ipld-prime/blob/adf99676901e35d42bba09c67a236ee0c87675f0/datamodel/node.go#L60. If you've got to add totally game to add them.

I think being able to work with chunks of a map asynchronously instead of via an iterator is a good idea.